Source/Particles/Deposition/CurrentDeposition.H

```diff
@@ -1533,6 +1533,670 @@ void doChargeConservingDepositionShapeNImplicit (const amrex::ParticleReal * con
+                                                     + sx_new_node[i]*sz_new_node[k]*one_third )*seg_factor_y;
+                            amrex::Gpu::Atomic::AddNoRet( &Jy_arr(lo.x+i0_node+i, lo.y+j0_cell+j, lo.z+k0_node+k), this_Jy);
+                        }
+                    }
+                }
+
+                // deposit Jz for this segment
+                amrex::Real this_Jz;
+                for (int i=0; i<=depos_order; i++) {
+                    for (int j=0; j<=depos_order; j++) {
+                        for (int k=0; k<=depos_order-1; k++) {
+                            this_Jz = wqz*sz_cell[k]*( sx_old_node[i]*sy_old_node[j]*one_third
+                                                     + sx_old_node[i]*sy_new_node[j]*one_sixth
+                                                     + sx_new_node[i]*sy_old_node[j]*one_sixth
+                                                     + sx_new_node[i]*sy_new_node[j]*one_third )*seg_factor_z;
+                            amrex::Gpu::Atomic::AddNoRet( &Jz_arr(lo.x+i0_node+i, lo.y+j0_node+j, lo.z+k0_cell+k), this_Jz);
+                        }
+                    }
+                }
+
+                // update old segment values
+                if (ns < num_segments-1) {
+                    x0_old = x0_new;
+                    y0_old = y0_new;
+                    z0_old = z0_new;
+                }
+
+            } // end loop over segments
+
+#elif defined(WARPX_DIM_XZ) || defined(WARPX_DIM_RZ)
+
+            // compute cell crossings in X-direction
+            const auto i_old = static_cast<int>(x_old-shift);
+            const auto i_new = static_cast<int>(x_new-shift);
+            int cell_crossings_x = std::abs(i_new-i_old);
+            num_segments += cell_crossings_x;
+
+            // compute cell crossings in Z-direction
+            const auto k_old = static_cast<int>(z_old-shift);
+            const auto k_new = static_cast<int>(z_new-shift);
+            int cell_crossings_z = std::abs(k_new-k_old);
+            num_segments += cell_crossings_z;
+
+            // need to assert that the number of cell crossings in each direction
+            // is within the range permitted by the number of guard cells
+            // e.g., if (num_segments > 3) ...
+
+            // compute total change in particle position and the initial cell
+            // locations in each direction used to find the position at cell crossings.
+            const double dxp = x_new - x_old;
+            const double dzp = z_new - z_old;
+            const auto dirX_sign = static_cast<int>(dxp < 0. ? -1 : 1);
+            const auto dirZ_sign = static_cast<int>(dzp < 0. ? -1 : 1);
+            double Xcell = 0., Zcell = 0.;
+            if (num_segments > 1) {
+                Xcell = i_old + shift + 0.5_rt*(1-dirX_sign);
+                Zcell = k_old + shift + 0.5_rt*(1-dirZ_sign);
+            }
```

Comment:
Idea/optional operator, but I would explicitly cast:
```suggestion
            double Xcell = num_segments > 1 ?
                static_cast<double>(i_old) + shift + 0.5*(1.0-dirX_sign) : 0.;
            double Zcell = num_segments > 1 ?
                static_cast<double>(k_old) + shift + 0.5*(1.0-dirZ_sign) : 0.;
```

Do you want to use `amrex::ParticleReal` (`_prt`) here, which can be `float` and `double`, or do you require `double` here?


Note that `shift` is currently defined as a `Real` (float or double, depending on compile).

Reply:
Will adopt the static cast. I'm still a little unclear on when to use the different types, but I think we should keep double here since it is used to find cell crossings and we want to minimize roundoff error.

Reply:
I will change shift to a double as well to be consistent

Reply:
By default we actually set `amrex::Real` and `amrex::ParticleReal` to `double`.